Fix backspace deleting 2 characters at a time

shiki opened the pull request:
Closes #5393 

## Findings

The Simple Payments numerical entry uses [`CurrencyEditText`](https://github.com/woocommerce/woocommerce-android/blob/develop/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt). This custom text field uses [regex filtering](https://github.com/woocommerce/woocommerce-android/blob/38223050ae60c8404e6eed90ae4497a5fd3ed144/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L62) to prevent the user from entering non-numeric characters. 

Based on my understanding, the backspace deleting 2 characters seem to be caused by these lines:

https://github.com/woocommerce/woocommerce-android/blob/38223050ae60c8404e6eed90ae4497a5fd3ed144/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L73-L75

And these lines were added in 2f71c15f586cfe4b602916f31a12d38f72ec459a which has a message "Fix the backspace edit in CurrencyEditText". But there doesn't seem to be enough context on what the problem is. (cc @0nko). 

## Solution

I removed the lines above and it seems to be working fine. 🤷 

I also added some unit tests for that function: [`CurrencyEditTestTest`](https://github.com/woocommerce/woocommerce-android/blob/issue/5393-fix-backspace/WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt). 

## Testing

1. Open the Simple Payments UI. 
2. Type 4 numbers. 
3. Press backspace. 
4. Confirm that only 1 number was deleted. 

### Regression

If I'm understanding Android Studio correctly, the `CurrencyEditText` is only used in Simple Payments and in Refund by Amount. The Refund by Amount has been disabled for a long time. IIRC, we're still waiting for an API change for that to be feasible. 

## Reviewing

Only 1 reviewer is needed but anyone can review.

Please feel free to nitpick me to the death. 🤪 It's been a while since I've written Kotlin. 

## Author Checklist

- [x] If it's feasible, I have added unit tests. 
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

---

cc @nbradbury

shiki committed:
CurrencyEditText: Move onTextChanged logic to a testable function

shiki committed:
Add unit tests for CurrencyEditText.clean()

One of these is failing and is what we'll be fixing next.

shiki committed:
Remove custom backspace behavior

This causes the backspace to delete 2 characters instead of just one. I'm not
entirely sure why we need this logic. But so far based on my tests, numeric
entry still seems to be working fine.

shiki committed:
Remove unused arguments

shiki committed:
Remove empty line

wzieba commented:
👋 I haven't got yet into code details but it makes backspace button unusable for my device. 

Also, I think that this issue happens only when there is 7 digits so I'd change instructions to `Type 7 numbers.`, is that correct? Because for less characters I'm not able to reproduce it. It's a little tricky so I'm still experimenting.

https://user-images.githubusercontent.com/5845095/145222739-4652432a-15e7-47d0-992d-2ecc386ea040.mov

shiki committed:
Revert "Remove unused arguments"

This reverts commit b70e53b9a6df263552661852657301454be55394.

shiki committed:
Revert "Remove custom backspace behavior"

This reverts commit e522dca2541ec50b9be5a04031402d787323ee72.

shiki committed:
Add failing unit tests for backspace keypress scenarios

shiki committed:
Change tests to indicate char deletion for symbols on the right

shiki committed:
Rename regex to nonNumericMatchPattern

shiki committed:
Only manually delete if the symbol is on the right

shiki committed:
Add comments

shiki committed:
Fix lint errors

shiki committed:
Fix lint warnings

shiki commented:
@wzieba Thank you very much for finding that problem. It looks like that code that I deleted disabled the backspace if the currency symbol is on the right. 🤦 

I was able to reproduce the bug that you showed. And when I restored the code in 41e5ff48baed03750fe72dbd9af8e397c2a1b6a3, it's fixed. 🎉 But it's broken if the symbol is on the left. 🤦 

Here's what I think is happening:

1. Assume that the current text is "$12.34"
2. User presses backspace
3. [`onTextChanged`](https://github.com/woocommerce/woocommerce-android/blob/cd655e7c1ae0b21e8cb558ca1dea174157b2e2b9/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L62) will be called with `text` = `$12.3`. Notice that the `4` was already deleted. 
4. Everything's great until this line: 

    https://github.com/woocommerce/woocommerce-android/blob/cd655e7c1ae0b21e8cb558ca1dea174157b2e2b9/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L73-L75
5. That code will be executed because `lengthBefore` is greater than `lengthAfter` because of the deletion. That code will pretty much delete the last digit. This is what is causing the double-digit deletion. At this point, the value will be `$12`. 

Now, if the symbol is on the **right**, this is what happens:

1. Assume that the current text is "12.34$" (symbol on the right)
2. User presses backspace
3. `onTextChanged` will be called with `text` = `12.34`. Notice that the `$` character was deleted.
4. The same [backspace handling line](https://github.com/woocommerce/woocommerce-android/blob/cd655e7c1ae0b21e8cb558ca1dea174157b2e2b9/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L73-L75) gets executed. But at this stage, it's okay because the result will be `12.3`. Notice `4` gets removed.

So I think the problem only happens if the currency symbol is on the left. I fixed this by adding this condition: 

https://github.com/woocommerce/woocommerce-android/blob/4da78b52433574ad37dbfb5da769ec78c0c2af19/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/CurrencyEditText.kt#L109-L113

We'll only do the **manual deletion** if there was (or is) a currency symbol on the right. It's not the best solution but it's what I came up with considering the current situation. 😅 

I've tested this as much as I can and it seems to be working fine. Would you mind testing and reviewing again? 🙏

wzieba reviewed (approved):
I've re-tested and now it works fine, thanks!

Thank you also for explanation! Let's merge this as it is and, as we talked on DM, we can revisit whole `CurrencyEditText` if it'll cause problems in the future.

0nko commented:
Thank you @shiki, @wzieba for addressing this. What an annoying bug!
